--- ixp400_eth.c_orig	2004-09-13 13:59:58.000000000 +0800
+++ ixp400_eth.c	2004-09-13 13:55:57.000000000 +0800
@@ -86,7 +86,7 @@
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Intel Corporation");
 #define MODULE_NAME "ixp400_eth"
-#define MODULE_VERSION "1.3"
+#define MODULE_VERSION "1.3A"
 
 /* Module parameters */
 static int npe_learning = 1; /* default : NPE learning & filtering enable */
@@ -780,6 +780,8 @@
 #ifdef CONFIG_NETFILTER
 #if defined(CONFIG_BRIDGE) || defined(CONFIG_BRIDGE_MODULE)
 /* We need to free the memory attached to the nf_bridge pointer to avoid a memory leak */
+	 nf_bridge_put(skb->nf_bridge);
+        skb->nf_bridge = NULL;

 #endif
 #endif
